feat(client): migrate MilvusClient to AsyncMilvusClient

Elbehery opened the pull request:
# What does this PR do?

This PR migrates `MilvusClient` to `AsyncMilvusClient`.

The commit makes the follwing changes.
-  Import statements updated: MilvusClient → AsyncMilvusClient
-  Removed asyncio.to_thread() wrappers: All Milvus operations now use native async/await
-  Test compatibility: Mock objects and fixtures updated to work with AsyncMilvusClient

<!-- If resolving an issue, uncomment and update the line below -->
<!-- Closes #[issue-number] -->

Closes https://github.com/llamastack/llama-stack/issues/2684

Commit:
feat(client): migrate MilvusClient to AsyncMilvusClient

The commit makes the follwing changes.
-  Import statements updated: MilvusClient → AsyncMilvusClient
-  Removed asyncio.to_thread() wrappers: All Milvus operations now use native async/await
-  Test compatibility: Mock objects and fixtures updated to work with AsyncMilvusClient

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

Commit:
fix(test): chrome db test fails due to reusing deleted collection

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

Commit:
chore: remove irrelevant comments

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

Commit:
fix(integration): init AsyncMilvusClient before MilvusIndex

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

mattf commented on llama_stack/providers/remote/vector_io/chroma/chroma.py:
```diff
@@ -106,7 +107,13 @@ async def query_vector(self, embedding: NDArray, k: int, score_threshold: float)
 
     async def delete(self):
-        await maybe_await(self.client.delete_collection(self.collection.name))
+        try:
```
@franciscojavierarceo ptal

franciscojavierarceo replied:
@Elbehery was that commit squashed? That link is not working for me.

franciscojavierarceo replied:
which test was failing?

Elbehery replied:
`Run the integration test suite with various VectorIO providers`

few runs 

- https://github.com/llamastack/llama-stack/actions/runs/17557595289
- https://github.com/llamastack/llama-stack/actions/runs/17556877844

search the log for 

```
[ERROR][has_collection]: Failed to check collection existence:
```

franciscojavierarceo replied:
hmm i don't see `has_collection()` is used in milvus, which i don't believe is available in Chroma (I tested the client locally and don't see it listed here: https://docs.trychroma.com/docs/collections/manage-collections).

Elbehery replied:
> has_collection() is used in milvus

yes that is what i meant. Therefore i have used the same method, and not `list_collection()` to minimize code changes.

> which i don't believe is available in Chroma

the chroma test that I cited above, started failing after the migration to `AsyncMilvusClient` 

Thus I have made these changes to fix the test in a separate [commit](https://github.com/llamastack/llama-stack/pull/3376/commits/aac26e8c6c2965fa6a5d2918c2b92ea34b074c46), so that it is easier for the reviewer.

Now I am trying use `list_collection()` in milvus client, and testing whether this would not fail  the `Chroma` test

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/chroma/chroma.py:
```diff
@@ -56,10 +56,11 @@ class ChromaIndex(EmbeddingIndex):
 
     async def initialize(self):
-        pass
+        self.collection = await maybe_await(self.client.get_or_create_collection(self.collection_name))
```
nice thank you.

Commit:
Revert "fix(test): chrome db test fails due to reusing deleted collection"

This reverts commit aac26e8c6c2965fa6a5d2918c2b92ea34b074c46.

Commit:
refactor(client): replace all AsyncMilvusClient usage of has_collection() with list_collections()

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -48,28 +47,58 @@
 
-        if not await asyncio.to_thread(self.client.has_collection, self.collection_name):
+        try:
+            collections = await self.client.list_collections()
```
given the usage here gain, it may make sense to just add a helper `_has_collection()` that has the try and except block and checks the `collection_name`

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -48,28 +47,58 @@
+            if "attached to a different loop" in str(e):
+                logger.warning("Recreating client due to event loop issue")
+
+                if hasattr(self, "_parent_adapter"):
```
why do we need this _parent_adapter? did you have issues with the event loop?

Elbehery replied:
yes, i had this issue 

```
 Problem: AsyncMilvusClient connections were established in one event loop but used in different event loops during test execution.

  # BROKEN - Client created in setup loop, used in test loop
  RuntimeError: Task <Task pending> got Future <Future pending> attached to a different loop

  Symptoms:
  - RuntimeError: attached to a different loop errors
  - Tests failing during both execution and teardown
  - Milvus operations hanging or crashing
```

the cause was that `MilvusIndex` which reference `MilvusClient` was being initialized _before_ `MilvusClient` 

Therefore the failure, the fix is to swap the init order, which is being fixed separately in this [commit](https://github.com/llamastack/llama-stack/pull/3376/commits/19f137436f6c5b2514c03c095770c32d42d70e25)

Elbehery replied:
yes, Claude

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -48,28 +47,58 @@
+        collection_name: str,
+        consistency_level="Strong",
+        kvstore: KVStore | None = None,
+        parent_adapter=None,
```
concerned about this.

Elbehery replied:
it is part of the fix of the init order in [commit](https://github.com/llamastack/llama-stack/pull/3376/commits/19f137436f6c5b2514c03c095770c32d42d70e25)

please let me know if there is a better approach :_

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -123,13 +152,16 @@ async def add_chunks(self, chunks: list[Chunk], embeddings: NDArray):
-                index_params=index_params,
-                consistency_level=self.consistency_level,
-            )
+            try:
```
nice thanks for adding this.

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -143,8 +175,7 @@ async def add_chunks(self, chunks: list[Chunk], embeddings: NDArray):
         try:
-            await asyncio.to_thread(
-                self.client.insert,
+            await self.client.insert(
```
probably we should improve the error in the raise

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -321,6 +346,15 @@ def __init__(
 
     async def initialize(self) -> None:
         self.kvstore = await kvstore_impl(self.config.kvstore)
+
```
it'd be good to also add exception handling here. i think someone had an issue with the db_path at some point and we probably could've caught that bug better since it can be configured in the run.yaml

franciscojavierarceo replied:
in this PR it's fine to add!

franciscojavierarceo replied:
i'm doing a little scope creep but while we're here it's a good thing to add a little more resilience as end users will be happy :D

Elbehery replied:
lgtm 💯 

will do tmrw, i am sleeping on my desk :D

franciscojavierarceo reviewed (changes requested):
i don't think we need `_recreate_client` can you provide more info on why this is necessary?

Elbehery commented:
> i don't think we need `_recreate_client` can you provide more info on why this is necessary?

yes, I saved all the failure of this, as I spent time fixing it and knew would get a lot of questions 

```
AsyncMilvusClient Migration Issue & Fix

  The Problem

  When migrating from the synchronous MilvusClient to AsyncMilvusClient, the codebase encountered two critical issues that caused widespread test failures:

  Issue 1: Client Initialization Order Bug

  Problem: The AsyncMilvusClient was being passed to MilvusIndex constructors before being initialized.

  # BROKEN CODE
  async def initialize(self) -> None:
      # Create indices with uninitialized client (client = None)
      for vector_db_data in stored_vector_dbs:
          index = MilvusIndex(
              client=self.client,  # ❌ self.client is None here!
              collection_name=vector_db.identifier,
          )

      # Client gets initialized AFTER being passed to indices
      self.client = AsyncMilvusClient(uri=uri)  # ❌ Too late!

  Symptoms:
  - AttributeError: 'NoneType' object has no attribute 'has_collection'
  - Index operations failing because they had no valid client

  Issue 2: AsyncIO Event Loop Issues

  Problem: AsyncMilvusClient connections were established in one event loop but used in different event loops during test execution.

  # BROKEN - Client created in setup loop, used in test loop
  RuntimeError: Task <Task pending> got Future <Future pending> attached to a different loop

  Symptoms:
  - RuntimeError: attached to a different loop errors
  - Tests failing during both execution and teardown
  - Milvus operations hanging or crashing

  The Solution

  Fix 1: Correct Initialization Order

  Solution: Initialize the client first, then create indices that reference it.

  # FIXED CODE
  async def initialize(self) -> None:
      # Initialize client FIRST
      if isinstance(self.config, RemoteMilvusVectorIOConfig):
          self.client = AsyncMilvusClient(**self.config.model_dump(exclude_none=True))
      else:
          uri = os.path.expanduser(self.config.db_path)
          self.client = AsyncMilvusClient(uri=uri)

      # Now create indices with valid client
      for vector_db_data in stored_vector_dbs:
          index = MilvusIndex(
              client=self.client,  # ✅ self.client is properly initialized
              collection_name=vector_db.identifier,
              parent_adapter=self,  # ✅ Added reference for client recreation
          )

  Fix 2: Event Loop Recovery Mechanism

  Solution: Detect event loop errors and automatically recreate the client.

  # ERROR DETECTION & RECOVERY
  async def add_chunks(self, chunks, embeddings):
      try:
          collection_exists = await self.client.has_collection(self.collection_name)
      except Exception as e:
          logger.error(f"Failed to check collection existence: {e}")

          # Detect event loop issues
          if "attached to a different loop" in str(e):
              logger.warning("Recreating client due to event loop issue")
              # Trigger client recreation
              if hasattr(self, '_parent_adapter'):
                  await self._parent_adapter._recreate_client()
                  collection_exists = await self.client.has_collection(self.collection_name)
              else:
                  collection_exists = False
          else:
              collection_exists = False

  Client Recreation Method:
  async def _recreate_client(self) -> None:
      """Recreate the AsyncMilvusClient when event loop issues occur"""
      try:
          # Close old client
          if self.client:
              await self.client.close()
      except Exception as e:
          logger.warning(f"Error closing old client: {e}")

      # Create new client in current event loop
      if isinstance(self.config, RemoteMilvusVectorIOConfig):
          self.client = AsyncMilvusClient(**self.config.model_dump(exclude_none=True))
      else:
          uri = os.path.expanduser(self.config.db_path)
          self.client = AsyncMilvusClient(uri=uri)

      # Update all cached indices to use new client
      for index_wrapper in self.cache.values():
          if hasattr(index_wrapper, 'index') and hasattr(index_wrapper.index, 'client'):
              index_wrapper.index.client = self.client

  Fix 3: Enhanced Error Handling

  Solution: Add graceful error handling throughout the codebase.

  # GRACEFUL TEARDOWN
  async def delete(self):
      try:
          if await self.client.has_collection(self.collection_name):
              await self.client.drop_collection(collection_name=self.collection_name)
      except Exception as e:
          logger.warning(f"Failed to check or delete collection {self.collection_name}: {e}")
          # Continue gracefully - collection might already be deleted or client disconnected

  Why This Approach Works

  1. Proactive Detection: The code detects event loop issues by examining exception messages
  2. Automatic Recovery: When detected, it automatically recreates the client in the current event loop
  3. Reference Updates: All cached indices get updated to use the new client
  4. Graceful Degradation: If recovery fails, operations continue with sensible defaults
  5. Clean Teardown: Tests can complete even when connections are in invalid states
```

Elbehery commented:
> What errors were you seeing that resulted in the alleged event loop issues? Can you provide the exact command you used to reproduce that?

after the migration, `Unit Tests` were failing, and I analyzed the failures, and it was due the order of the init

> It would be better to reproduce those errors in the CI. 

It was caught by the CI tests 👍🏽 

> Probably you could do that by removing the _recreate_client changes.

I am searching the logs of Claude for more context

Elbehery commented:
> Initializing the MilvusClient before the index makes sense and we should keep that.
> 
> What errors were you seeing that resulted in the alleged event loop issues? Can you provide the exact command you used to reproduce that?
> 
> It would be better to reproduce those errors in the CI. Probably you could do that by removing the `_recreate_client` changes.

there are all i could find 

```
 1. Lines 358-375: The code loops through stored_vector_dbs and creates MilvusIndex objects, passing self.client as the client parameter
  2. Lines 350-356: But self.client is only initialized after the loop that creates the indices

  Here's the problematic flow:

  async def initialize(self) -> None:
      self.kvstore = await kvstore_impl(self.config.kvstore)

      # ❌ Client is still None here when creating indices
      start_key = VECTOR_DBS_PREFIX
      end_key = f"{VECTOR_DBS_PREFIX}\xff"
      stored_vector_dbs = await self.kvstore.values_in_range(start_key, end_key)

      for vector_db_data in stored_vector_dbs:
          vector_db = VectorDB.model_validate_json(vector_db_data)
          index = VectorDBWithIndex(
              vector_db,
              index=MilvusIndex(
                  client=self.client,  # ❌ self.client is None here!
                  collection_name=vector_db.identifier,
                  # ...
              ),
              # ...
          )
          self.cache[vector_db.identifier] = index

      # ✅ Client gets initialized AFTER being passed to indices  
      if isinstance(self.config, RemoteMilvusVectorIOConfig):
          logger.info(f"Connecting to Milvus server at {self.config.uri}")
          self.client = AsyncMilvusClient(**self.config.model_dump(exclude_none=True))
      else:
          logger.info(f"Connecting to Milvus Lite at: {self.config.db_path}")
          uri = os.path.expanduser(self.config.db_path)
          self.client = AsyncMilvusClient(uri=uri)

  The fix is to move the client initialization before the loop that creates the indices. The client initialization code (lines 350-356) should come before lines 358-375.
```

but i pasted [here](https://github.com/llamastack/llama-stack/pull/3376#issuecomment-3272183631) the actual context

Commit:
fix: resolve rebase conflicts

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

Commit:
fix(unittest): add pymilvus and milvus-lite to unit dep group

This commit resolves unittest-3.12 issues.

- It adds pymilvus and milvus-lite to unit dep group
- It rename VectorDB to VectorStore

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>

Commit:
fix(unittest): add required chunk_id field to Milvus test mock data

  The Chunk Pydantic model requires chunk_id as a mandatory field, but the
  mock data in test_milvus.py was missing this field in chunk_content objects,
  causing 6 test failures with ValidationError.

Signed-off-by: Mustafa Elbehery <melbeher@redhat.com>
